### PR TITLE
Add DLNA support

### PR DESCRIPTION
**Pull request by grunjol**
This PR adds DLNA support, fixing #186

### CODE REVIEW

File: renderer/lib/cast.js [hunk 1]
```diff
@@ -37,7 +38,7 @@ function chromecastPlayer (player) {
     var torrentSummary = state.saved.torrents.find((x) => x.infoHash === state.playing.infoHash)
     player.play(state.server.networkURL, {
       type: 'video/mp4',
-      title: config.APP_NAME + ' — ' + torrentSummary.name
+      title: config.APP_NAME + ' - ' + torrentSummary.name
```
Comment (feross) [inline renderer/lib/cast.js]: Could DLNA not handle the wide em dash character?

File: renderer/index.js [hunk 2]
```diff
@@ -241,11 +241,8 @@ function dispatch (action, ...args) {
   if (action === 'openTorrentContextMenu') {
     openTorrentContextMenu(args[0] /* infoHash */)
   }
-  if (action === 'openChromecast') {
-    lazyLoadCast().open('chromecast')
-  }
-  if (action === 'openAirplay') {
-    lazyLoadCast().open('airplay')
+  if (action === 'open') {
```
Comment (feross) [inline renderer/index.js]: This name is too generic. How about `openDevice`?

File: renderer/views/player.js [hunk 3]
```diff
@@ -275,13 +290,22 @@ function renderPlayerControls (state) {
...
         airplay
       </i>
     `)
   }
+  if (state.devices.dlna || isOnDlna) {
+    elements.push(hx`
+      <i.icon.device
+        class=${dlnaClass}
+        onclick=${dlnaHandler}>
+        cast
```
Comment (feross) [inline renderer/views/player.js]: Let's not use the same icon as Chromecast. Can you change this to `tv`?

File: renderer/views/player.js [hunk 4]
```diff
@@ -181,21 +181,22 @@ function renderLoadingSpinner (state) {
...
-  if (!isChromecast && !isAirplay) throw new Error('Unimplemented cast type')
+  if (!isChromecast && !isAirplay && !isDlna) throw new Error('Unimplemented cast type')
 
   // Show a nice title image, if possible
   var style = {
     backgroundImage: cssBackgroundImagePoster(state)
   }
 
-  // Show whether we're connected to Chromecast / Airplay
+  // Show whether we're connected to Chromecast / Airplay /DLNA
```
Comment (feross) [inline renderer/views/player.js]: Space after the slash for consistent spacing

File: renderer/index.css [hunk 5]
```diff
@@ -668,8 +668,7 @@ body.drag .app::after {
   margin: 0 auto;
 }
 
-.player-controls .chromecast,
-.player-controls .airplay,
+.player-controls .device,
```
Comment (feross) [inline renderer/index.css]: Good call, this is better than continuing to add classes for each device type.

File: package.json [hunk 6]
```diff
@@ -16,6 +16,7 @@
     "application-config-path": "^0.1.0",
     "chromecasts": "^1.8.0",
     "create-torrent": "^3.22.1",
+    "dlnacasts": "^0.0.1",
```
Comment (feross) [inline package.json]: Nice work creating this!

**Comment by feross** [comment]
What is a cheap DLNA device that I can buy to test this out?

**Comment by feross** [comment]
Also, before we merge this, can you get rid of the `async` dependency in `dlnacasts`?

`async` is bloated and includes 50+ useless functions, plus all of `lodash`. I've never needed more than a couple async functions, just `parallel` and `series` are adequate 99% of the time. Can you swap out your `async.parallel` call with [`run-parallel`](https://www.npmjs.com/package/run-parallel) please?

Cheers!

@dcposch Can you review this too?

File: renderer/lib/cast.js (hunk 1 above)
Comment (grunjol) [inline renderer/lib/cast.js]: its funny, for some weird reason the long dash breaks the DLNA service, i think I copied it back here to create consistency with the other call (maybe reuse it in the future?)

File: renderer/index.js (hunk 2 above)
Comment (grunjol) [inline renderer/index.js]: good

File: package.json (hunk 6 above)
Comment (grunjol) [inline package.json]: I didn't find any complete solution with a good API, so i made it based on chromecasts

**Comment by grunjol** [comment]
Cheap DLNA device? any cheap android TV stick can handle DLNA these days. 
Even more, if to have a Samsung or Phillips 'smart tv' ia m pretty sure it supports it.
You can even install a DLNA renderer in linux (`apt-get install gmediarender`) in virtualbox

**Comment by grunjol** [comment]
Will swap async with run-parallel and made all the changes you suggested

File: package.json (hunk 6 above)
Comment (feross) [inline package.json]: Nice :)

File: renderer/lib/cast.js (hunk 1 above)
Comment (feross) [inline renderer/lib/cast.js]: Okay, consistency is a good reason.

**Comment by feross** [comment]
Cool, will test out on a Samsung Smart TV later tonight.

Will merge after you've made the changes.

**Comment by grunjol** [comment]
changes updated

**Comment by feross** [comment]
Okay, I just had a chance to test this out with an Xbox.

It worked for an `.mp4` file, but not a `.webm` file. Nothing we can do about that if the Xbox doesn't support .webm, I suppose.

But we need better error handling. We need to let the user know that the streaming failed and give a reason, if we can.

I'll merge this now, but let's try to improve it over the coming days and weeks.

**Comment by grunjol** [comment]
I didn't catch it because in my tests with android tv sticks (2) and TVs (2), webm was supported (all of them are >2014). 

@feross: completely agree with you, we need to let user know something went wrong.
In fact, its not DLNA specific. If chromcast fail to connect, it silently fails and revert back to internal player. In airplay case, its added to state.errors
```
        state.errors.push({
          time: new Date().getTime(),
          message: 'Couldn\'t connect to Airplay'
        })
```
but not notified to the user neither.

I can catch the unsupported media error and notify the user, leading me to the question. Is there a notification system in webtorrent-desktop? I was thinking something like Android's [Toast](http://developer.android.com/guide/topics/ui/notifiers/toasts.html), or more complex, use some kind of native notification, but i think this short, program related notifications should be handled locally.

**Comment by grunjol** [comment]
today i just need to read the code and not asking questions, status.error was the native solution.
Updating PR.

**Comment by feross** [comment]
> I didn't catch it because in my tests with android tv sticks (2) and TVs (2), webm was supported (all of them are >2014).

It's great that most devices support `.webm` these days.

> If chromecast fail to connect, it silently fails and revert back to internal player. 

I'll fix that in a PR shortly.

The rest of your changes look good.

**Comment by feross** [comment]
@dcposch On second thought, I think the changes in this PR are fine to merge before the release :)

**Comment by feross** [comment]
@grunjol Thanks for this, and all your other, great PRs! I just added you as a GitHub collaborator. Congrats!

Please read the [contributing guidelines](https://github.com/feross/webtorrent-desktop/blob/master/CONTRIBUTING.md) so you know what responsibilities collaborators have. Cheers! :beers:

File: renderer/views/player.js [hunk 7]
```diff
@@ -181,21 +181,22 @@ function renderLoadingSpinner (state) {
...
     backgroundImage: cssBackgroundImagePoster(state)
   }
 
-  // Show whether we're connected to Chromecast / Airplay
+  // Show whether we're connected to Chromecast / Airplay / DLNA
   var castStatus = isStarting ? 'Connecting...' : 'Connected'
   return hx`
     <div class='letterbox' style=${style}>
       <div class='cast-screen'>
         <i class='icon'>${isAirplay ? 'airplay' : 'cast'}</i>
```
Comment (dcposch) [inline renderer/views/player.js]: This means you can have two identical looking cast icons, one for Chromecast and one for Airplay, right?

**Comment by dcposch** [comment]
@feross this broke all casting as well as seeking for me. every time `lazyLoadCasts()` runs, it throws this error:

```
module.js:338 Uncaught Error: Cannot find module 'mime'
```

**Comment by feross** [comment]
No chooser yet, but I think that's fine.

Looking at the mime error.

**Comment by feross** [comment]
Oh, I see what happened. `dlnacasts` doesn't have `mime` in the `package.json`, but `mime` got hoisted to the top level by `npm dedupe` for me, which is why I didn't notice it.

**Comment by feross** [comment]
@dcposch Fixed on master.

**Comment by grunjol** [comment]
fixed in dlnacasts 0.0.3

**Pull request merged**